src/transformers/models/bloom/parallel_layers.py

```diff
@@ -0,0 +1,235 @@
+
+        # return F.linear(input, weight=weight.transpose(1,0), bias=bias)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        out = self.linear(input, weight=self.weight, bias=self.bias)
+        torch.distributed.all_reduce(out, group=self.process_group)
+
+        # ### DEBUG @thomasw21:: Check that shard model output the same as the non sharded version
+        # sharded_out = out
+        #
+        # input_from_tp_ranks = [torch.empty_like(input) for _ in range(self.process_group.size())]
+        # weight_from_tp_ranks = [torch.empty_like(self.weight) for _ in range(self.process_group.size())]
+        # bias = self.bias.clone()
+        # torch.distributed.all_gather(input_from_tp_ranks, input, group=self.process_group)
+        # torch.distributed.all_gather(weight_from_tp_ranks, self.weight, group=self.process_group)
+        # torch.distributed.all_reduce(bias, group=self.process_group)
+        # input = torch.cat(input_from_tp_ranks, dim=-1)
+        # weight = torch.cat(weight_from_tp_ranks, dim=1)
+        # baseline_out = F.linear(input, weight, bias)
+        #
+        # if self.process_group.rank() == 0:
+        #     torch.testing.assert_close(bias, self.bias, atol=0.0, rtol=0.0)
+        # torch.distributed.barrier(self.process_group)
+        # # torch.testing.assert_close(sharded_out, baseline_out)
+        # ###
+
+        return out
+
+    def extra_repr(self) -> str:
+        """From `torch.nn.Linear`"""
+        return 'in_features={}, out_features={}, bias={}'.format(
+            self.in_features, self.out_features, self.bias is not None
+        )
+
+class TensorParallelEmbedding(nn.Embedding):
+    def __init__(
+        self,
+        num_embeddings,
+        embedding_dim,
+        process_group: torch.distributed.ProcessGroup,
+        padding_idx=None,
+        max_norm=None,
+        norm_type=2.0,
+        scale_grad_by_freq=False,
+        sparse=False,
+        _weight=None,
+        device=None,
+        dtype=None
+    ):
+        self.process_group = process_group
+        self.tp_rank = process_group.rank()
+        self.tp_world_size = process_group.size()
+
+        self.original_num_embeddings = num_embeddings
+
+        # TODO @thomasw21 fix and remove that constraint
+        assert num_embeddings % self.tp_world_size == 0
+        block_size = num_embeddings // self.tp_world_size
+        # inputs in `[min_id, max_id[` are handled by `self` to get embeddings
+        self.min_id = self.tp_rank * block_size
+        self.max_id = (self.tp_rank + 1) * block_size
+
+        super().__init__(block_size, embedding_dim, padding_idx=padding_idx, max_norm=max_norm, norm_type=norm_type, scale_grad_by_freq=scale_grad_by_freq, sparse=sparse, _weight=_weight, device=device, dtype=dtype)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        # Sanity check
+        if torch.any(torch.logical_or(0 > input, input >= self.original_num_embeddings)):
```

Comment:
Interestingly this can be quite expensive, it's a sanity check that maybe we should be able to deactivate.

Reply:
Commenteed out.